### PR TITLE
predict emag and make blessed record struct events

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR

- moves all of emag into shared.
- use timespans instead of accumulators
- support pausing
- access and networking
- prediction on the event itself

**Media**

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

no cl no fun

### CODE REVIEW

File: Content.Shared/Emag/Systems/EmagSystem.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                _popupSystem.PopupEntity(Loc.GetString("emag-no-charges"), user, user);
+                return false;
+            }
+
+            var handled = DoEmagEffect(user, target);
+            if (!handled)
+                return false;
+
+            // only do popup on client
+            if (_net.IsClient && _timing.IsFirstTimePredicted)
```
Comment (metalgearsloth) [inline Content.Shared/Emag/Systems/EmagSystem.cs:142]: is IsFirstTimePredicted true on server?

**Review by metalgearsloth — APPROVED** [review]

File: Content.Shared/Emag/Systems/EmagSystem.cs (hunk 1 above)
Comment (EmoGarbage404) [reply Content.Shared/Emag/Systems/EmagSystem.cs:142]: does it matter, considering it only runs on the client?

Comment (metalgearsloth) [reply Content.Shared/Emag/Systems/EmagSystem.cs:142]: Was wondering whether .IsClient was necessary.

Comment (EmoGarbage404) [reply Content.Shared/Emag/Systems/EmagSystem.cs:142]: pretty sure it does.

**Pull request merged**